src/rust-crypto/verification.ts

```diff
@@ -0,0 +1,429 @@
+     */
+    public getQRCodeBytes(): Buffer | undefined {
+        // TODO
+        return undefined;
+    }
+
+    /**
+     * If this request has been cancelled, the cancellation code (e.g `m.user`) which is responsible for cancelling
+     * this verification.
+     */
+    public get cancellationCode(): string | null {
+        throw new Error("not implemented");
+    }
+
+    /**
+     * The id of the user that cancelled the request.
+     *
+     * Only defined when phase is Cancelled
+     */
+    public get cancellingUserId(): string | undefined {
+        throw new Error("not implemented");
+    }
+}
+
+export class RustSASVerifier extends TypedEventEmitter<VerifierEvent, VerifierEventHandlerMap> implements Verifier {
+    /** A promise which completes when the verification completes (or rejects when it is cancelled/fails) */
+    private readonly completionPromise: Promise<void>;
+
+    private callbacks: ShowSasCallbacks | null = null;
+
+    public constructor(
+        private readonly inner: RustSdkCryptoJs.Sas,
+        _verificationRequest: RustVerificationRequest,
+        private readonly outgoingRequestProcessor: OutgoingRequestProcessor,
+    ) {
+        super();
+
+        this.completionPromise = new Promise<void>((resolve, reject) => {
+            const onChange = async (): Promise<void> => {
+                this.updateCallbacks();
+
+                if (this.inner.isDone()) {
+                    resolve(undefined);
+                } else if (this.inner.isCancelled()) {
+                    const cancelInfo = this.inner.cancelInfo()!;
+                    reject(
+                        new Error(
+                            `Verification cancelled by ${
+                                cancelInfo.cancelledbyUs() ? "us" : "them"
+                            } with code ${cancelInfo.cancelCode()}: ${cancelInfo.reason()}`,
+                        ),
+                    );
+                }
+            };
+            inner.registerChangesCallback(onChange);
+        });
+        // stop the runtime complaining if nobody catches a failure
+        this.completionPromise.catch(() => null);
+    }
+
+    /** if we can now show the callbacks, do so */
+    private updateCallbacks(): void {
+        if (this.callbacks === null) {
+            const emoji: Array<Emoji> | undefined = this.inner.emoji();
+            const decimal = this.inner.decimals() as [number, number, number] | undefined;
+
+            if (emoji === undefined && decimal === undefined) {
+                return;
+            }
+
+            this.callbacks = {
+                sas: {
+                    decimal: decimal,
+                    emoji: emoji?.map((e) => [e.symbol, e.description]),
+                },
+                confirm: async (): Promise<void> => {
+                    const requests: Array<OutgoingRequest> = await this.inner.confirm();
+                    for (const m of requests) {
```

Comment:
Maybe a bone question. Why `m`?

Reply:
erm, good question. m for message, maybe? Though again, #3490 rather than new in this PR.